uniffi_bindgen/src/bindings/swift/templates/Helpers.swift

```diff
@@ -0,0 +1,109 @@
+
+
+{% if ci.iter_callback_interface_definitions().len() > 0 %}
+
+class Lock {
+    init() {}
+    let lock = NSLock()
+    func withLock<T>(f: () throws -> T) rethrows -> T {
+        lock.lock()
+        defer { lock.unlock() }
+        return try f()
+    }
+}
+
+typealias Handle = UInt64
+class ConcurrentHandleMap<T: AnyObject> {
+    private var leftMap: [Handle: T] = [:]
+    private var counter: [Handle: UInt64] = [:]
+    private var rightMap: [ObjectIdentifier: Handle] = [:]
+
+    private let lock = Lock()
+    private var currentHandle: Handle = 0
+    private let stride: Handle = 1
+
+    func insert(obj: T) -> Handle {
+        lock.withLock {
+            let handle = rightMap[ObjectIdentifier(obj)] ?? {
+                currentHandle += stride
+                let handle = currentHandle
+                leftMap[handle] = obj
+                rightMap[ObjectIdentifier(obj)] = handle
+                return handle
+            }()
+            counter[handle] = (counter[handle] ?? 0) + 1
+            return handle
+        }
+    }
+
+    func callWithResult<R>(handle: Handle, fn: (T) -> R) -> R {
+        guard let obj = lock.withLock(f: { leftMap[handle] })
+        else { fatalError("Panic: handle not in handlemap") }
+        return fn(obj)
+    }
+
+    func get(handle: Handle) -> T? {
+        lock.withLock {
+            leftMap[handle]
+        }
+    }
+
+    func delete(handle: Handle) {
+        remove(handle: handle)
+    }
+
+    @discardableResult
+    func remove(handle: Handle) -> T? {
+        lock.withLock {
+            defer { counter[handle] = (counter[handle] ?? 1) - 1 }
```

Comment:
I was confused about mention of ref counting here. 

I'm going to think outloud, without the refcounting you're doing:

1. If I pass the same callback in to Rust twice, and Rust creates a "shadow" implementations for each
2. Rust `drop`s one of the shadows, causing Swift to remove the actual callback interface
3. Rust makes a call on the other shadow, and Swift `fatalError`s?.

Would this be ameliorated if you were passing around/storing in the handlemap different instances of `{{ type_name }}Erased` holding the same callback interface ?

Reply:
That's really smart, let swift do the counting for us! That means we can remove the guard in insert
let handle = rightMap[ObjectIdentifier(obj)] ?? {
and that means there are no reads to rightMap so that can be removed too!
I tested a little and it seems to work!
If you think this is right I'll get rid of the counting and the right map.

Reply:
Replying to the part about implementing ViaFfi on the generated callback types it seems like it might not be possible. I agree that it would be really nice to conform to the protocol, but since we generate the code its not really needed. I find solving these types of problems in swift are really hard since they explore parts of the language that are not normally used. Its like solving a puzzle that might not have a solution.

Reply:
> I find solving these types of problems in swift are really hard since they explore parts of the language that are not normally used. Its like solving a puzzle that might not have a solution.

This is very nicely put. And we're trying to solve parallel interlocking puzzles in multiple languages. That may not have solutions. :)